mipengine/node/monetdb_interface/common_actions.py

```diff
@@ -408,49 +407,72 @@ def _convert_mip2monet_table_type(table_type: TableType) -> int:
 
 
-@sql_injection_guard(table_type=None, context_id=str.isalnum)
-def _drop_table_by_type_and_context_id(table_type: TableType, context_id: str):
+@sql_injection_guard(context_id=str.isalnum)
+def _get_tables_by_type(context_id: str) -> Dict[TableType, List[str]]:
     """
-    Drops all tables of specific type with name that contain a specific context_id from the DB.
+    Retrieve table names by type that contains a specific context_id from the DB.
 
     Parameters
     ----------
-    table_type : str
-        The type of the table
     context_id : str
         The id of the experiment
     """
     table_names_and_types = db_execute_and_fetchall(
         f"""
         SELECT name, type FROM tables
         WHERE name LIKE '%{context_id.lower()}%'
-        AND tables.type = {str(_convert_mip2monet_table_type(table_type))}
         AND system = false
         """
     )
+    table_names_by_type = {}
     for name, table_type in table_names_and_types:
-        if table_type == _convert_mip2monet_table_type(TableType.VIEW):
-            db_execute(f"DROP VIEW {name}")
-        else:
-            db_execute(f"DROP TABLE {name}")
+        mip_table_type = _convert_monet2mip_table_type(table_type)
+        if mip_table_type not in table_names_by_type:
+            table_names_by_type[mip_table_type] = []
+        table_names_by_type[mip_table_type].append(name)
+    return table_names_by_type
 
 
 @sql_injection_guard(context_id=str.isalnum)
-def _drop_udfs_by_context_id(context_id: str):
+def _get_function_names_query_by_context_id(context_id: str) -> List[str]:
     """
-    Drops all functions of specific context_id from the DB.
+    Retrieve all functions of specific context_id from the DB.
 
     Parameters
     ----------
     context_id : str
         The id of the experiment
     """
-    function_names = db_execute_and_fetchall(
+    result = db_execute_and_fetchall(
         f"""
         SELECT name FROM functions
         WHERE name LIKE '%{context_id.lower()}%'
         AND system = false
         """
     )
-    for name in function_names:
-        db_execute(f"DROP FUNCTION {name[0]}")
+    return [attributes[0] for attributes in result]
+
+
+def _get_drop_udfs_query(function_names: str):
+    return "".join([f"DROP FUNCTION {name};" for name in function_names])
+
+
+def _get_drop_tables_query(table_names_by_type: Dict[TableType, List[str]]):
+    # Order of the table types matter not to have dependencies when dropping the tables
+    table_type_drop_order = (
+        TableType.MERGE,
+        TableType.REMOTE,
+        TableType.VIEW,
+        TableType.NORMAL,
+    )
+    table_deletion_query = "".join(
+        [
+            f"DROP VIEW {name};"
+            if table_type == TableType.VIEW
+            else f"DROP TABLE {name};"
+            for table_type in table_type_drop_order
+            if table_type in table_names_by_type
+            for name in table_names_by_type[table_type]
+        ]
+    )
```

Comment:
From what was discussed with Yannis Foufoulas I thought that the DROP statements were supposed to lie within a single transaction. Is this wrong?